### ISSUE TITLE
[FEATURE] Support for RegexCommandHandler (regex command handler)

### ISSUE THREAD

**Issue by anton-samokat** [python-telegram-bot/python-telegram-bot#3406]
### What kind of feature are you missing? Where do you notice a shortcoming of PTB?

To use it like this:
`dispatcher.add_handler(RegexCommandHandler(["params_info.*", "pi.*"], show_params_info_cmd))`

### Describe the solution you'd like

Example with working implementation:

```python
import re
import warnings
from typing import Callable, Dict, List, Optional, Tuple, TypeVar, Union

from telegram import MessageEntity, Update
from telegram.ext import BaseFilter, Filters, CommandHandler
from telegram.ext.utils.types import CCT
from telegram.utils.deprecate import TelegramDeprecationWarning
from telegram.utils.types import SLT
from telegram.utils.helpers import DefaultValue, DEFAULT_FALSE

from samokat.common.logger_utils import config_logger, DEFAULT_LOGGING_LEVEL

RT = TypeVar('RT')

logger = config_logger('regexp_command_handler')
logger.setLevel(DEFAULT_LOGGING_LEVEL)


class RegexCommandHandler(CommandHandler[Update, CCT]):

    # __slots__ = ('commands_regexps', 'filters', 'pass_args')

    def __init__(
        self,
        command_regex: SLT[str],
        callback: Callable[[Update, CCT], RT],
        filters: BaseFilter = None,
        allow_edited: bool = None,
        pass_args: bool = False,
        pass_update_queue: bool = False,
        pass_job_queue: bool = False,
        pass_user_data: bool = False,
        pass_chat_data: bool = False,
        run_async: Union[bool, DefaultValue] = DEFAULT_FALSE,
    ):
        super().__init__(
            command="mock_command_to_path_check_in_parent_class",
            callback=callback,
            filters=filters,
            allow_edited=allow_edited,
            pass_args=pass_args,
            pass_update_queue=pass_update_queue,
            pass_job_queue=pass_job_queue,
            pass_user_data=pass_user_data,
            pass_chat_data=pass_chat_data,
            run_async=run_async,
        )

        if isinstance(command_regex, str):
            self.commands_regexps = [command_regex.lower()]
        else:
            self.commands_regexps = [x.lower() for x in command_regex]
        # anton-samokat - regex support start
        # todo make proper commands_regexps validation
        # for comm in self.command:
        #     if not re.match(r'^[\da-z_]{1,32}$', comm):
        #         raise ValueError('Command is not a valid bot command')
        # regex support end

        # todo below code from super class needed here?
        if filters:
            self.filters = Filters.update.messages & filters
        else:
            self.filters = Filters.update.messages

        if allow_edited is not None:
            warnings.warn(
                'allow_edited is deprecated. See '
                'https://github.com/python-telegram-bot/python-telegram-bot/wiki/Transition'
                '-guide-to-Version-12.0 for more info',
                TelegramDeprecationWarning,
                stacklevel=2,
            )
            if not allow_edited:
                self.filters &= ~Filters.update.edited_message
        self.pass_args = pass_args

    def check_update(
        self, update: object
    ) -> Optional[Union[bool, Tuple[List[str], Optional[Union[bool, Dict]]]]]:
        """Determines whether an update should be passed to this handlers :attr:`callback`.

        Args:
            update (:class:`telegram.Update` | :obj:`object`): Incoming update.

        Returns:
            :obj:`list`: The list of args for the handler.

        """
        if isinstance(update, Update) and update.effective_message:
            message = update.effective_message

            if (
                message.entities
                and message.entities[0].type == MessageEntity.BOT_COMMAND
                and message.entities[0].offset == 0
                and message.text
                and message.bot
            ):
                command = message.text[1: message.entities[0].length]
                args = message.text.split()[1:]
                command_parts = command.split('@')
                command_parts.append(message.bot.username)

                # anton-samokat - regex support start
                if not (
                    # command_parts[0].lower() in self.command
                    command_parts[1].lower() == message.bot.username.lower()
                ):
                    return None

                received_command = command_parts[0].lower()
                command_matched_regexps = False
                for regexp_pattern in self.commands_regexps:
                    pattern = re.compile(regexp_pattern)
                    if re.match(pattern, received_command):
                        logger.info(f"Command: {received_command}, matched regex: {regexp_pattern}")
                        command_matched_regexps = True
                        break

                if not command_matched_regexps:
                    return None
                # regex support end

                filter_result = self.filters(update)
                if filter_result:
                    return args, filter_result
                return False
        return None
```

**Comment by Bibo-Joshi** [comment]
Hi. Can you please explain in a few sentences what your proposed handler class can do that already existing handlers can not? Please also not that your implementation seems to be based on ptb v13, which no longer is under active development.

**Comment by anton-samokat** [comment]
@Bibo-Joshi Right, this implementation is for ptb v13. This command handler can handle commands based on `regex`. For example this:
`dispatcher.add_handler(RegexCommandHandler(["params_info.*", "pi.*"], show_params_info_cmd))`
handler will be taking and handle the following commands:
- params_info
- params_info__some_param
- params_info__another_param
etc.

I have not found how to implement such behavior with existing handlers. Is it possible with single handler? Instead of several command handlers like this:
```
dispatcher.add_handler(CommandHandler(["params_info"], show_params_info_cmd))
dispatcher.add_handler(CommandHandler(["params_info__some_param"], show_params_info_cmd))
dispatcher.add_handler(CommandHandler(["params_info__another_param"], show_params_info_cmd))
```

**Comment by Bibo-Joshi** [comment]
I see two possible options that already work:

1. For a fixed list of commands that should all trigger the same callback, you can use `CommandHanler(["params_info", "params_onf__some_param", "params_info__another_param"], show_params_info_cmd)`. This is ofc not regex based, but one can also
2. Combine `Filters.regex` and `Filters.command`: `MessageHandler(Filters.command & Filters.regex("params_info.*"), show_params_info_cmd)`

At first glance I don't see how how `RegexCommandHandler` would cover anything that the second option doesn't cover …

**Comment by anton-samokat** [comment]
@Bibo-Joshi 
I checked the second option. Yes, `Filters.regex` in conjunction with `MessageHandler` is really cover this case. `RegexCommandHandler` is not needed.

I noticed a small difference for command handler and message handler. For example, for input `/params_info param1 param2`:
- In `CommandHandler` `context.args` parameter is filled with `param1`, `param2` args.
- In `MessageHandler(Filters.command & Filters.regex("params_info.*"), show_params_info_cmd)` `context.args` is `None`. But this is not a big problem for user code. Can be parsed from `cmd` or `user_input` in user callback (`show_params_info_cmd` in our example). 

Still probably may be such `context.args` filling also makes sense to add to `MessageHandler` if it has `Filters.command` filter.

Current FR can be closed. Thanks.
